iroh-rpc/src/stream.rs

```diff
@@ -1,96 +1,190 @@
+                        chunks,
+                        task,
+                        error_sender,
+                        shutdown_receiver,
+                    )
+                    .await
+                })
+            },
+            early_shutdown: Some(shutdown_sender),
         }
     }
+
+    // we have loop that is taking in packets and ordering them. If there is an
+    // error, send the error to the main stream
+    async fn gather_packets(
+        mut packet_receiver: StreamReceiver,
+        chunks: Arc<Mutex<HashMap<u64, Vec<u8>>>>,
+        task: Arc<Mutex<AtomicWaker>>,
+        mut error_sender: mpsc::Sender<RpcError>,
+        mut shutdown: oneshot::Receiver<()>,
+    ) {
+        loop {
+            select! {
+              res = packet_receiver.next() => {
+                if let Some(r) = res {
+                  match r {
+                    StreamType::Packet(p) => {
+                      let mut chunks = chunks.lock().expect("failed to lock stream chunk mutex");
+                      chunks.insert(p.index, p.data);
+                      debug!(target="OrderedStream", "stream received chunk {}", p.index);
+                  },
+                    StreamType::RpcError(e) => {
+                      error!(target="OrderedStream", "stream received error {}", e);
+                      error_sender
+                          .send(e)
+                          .await
+                          .expect("failed to send on error sender");
+                    },
+                  }
+                  // notify poll_next that we have a message we want them to
+                  // process
+                  let task = task.lock().expect("failed to lock task");
+                  task.wake();
+                  continue;
+                }
+                // if res == None, then the packet_receiver stream has closed
+                    debug!(target="OrderedStream", "packet receiver closed, done
+                      gathering packets");
+                      return;
+                                },
+              _ = shutdown => {
+                debug!(target="OrderedStream", "early shutdown triggered, no longer
+                  gathering packets");
+                error_sender.send(RpcError::StreamClosedEarly).await.expect("failed to send on
+                  error_sender");
+                let task = task.lock().expect("failed to lock task");
+                task.wake();
+                return;
+              }
+            }
+        }
+    }
+
+    /// close cleans up the stream. MUST BE CALLED AFTER YOU ARE FINISHED WITH THE STREAM.
+    pub async fn close(mut self) {
+        self.out_sender
+            .try_send(Command::CloseStream { id: self.id })
+            .expect("Out sender to still be active.");
+        if let Some(sender) = self.early_shutdown.take() {
+            debug!(target = "OrderedStream", "closing stream");
+            self.early_shutdown = None;
```

Comment:
the call to `.take()` before already does this

Reply:
ah thank you, did a bad refactor... I'm relying too much on the compiler yelling at me.